### ISSUE TITLE
Add fully-blown support for Genycloud emulators

### ISSUE THREAD

**Issue by d4vidi** [wix/Detox#2655]
A follow-up on the work on Genymotion-cloud integration done in the context of a grand PoC in #2429.
While #2429 sums up the work required in order to start rolling out Genymotion cloud support internally at Wix, there is more to be done in case the integration proves scalable and robust - so as to make it a polished, generally used solution provided via Detox. Here is the list:

- [ ] [wip] Better tweaking of Detox' CI script, so as to support exec on both local and cloud emulators in each build.
- [ ] Documentation + scaffolding script revisit. **Important:** must cover the global Detox init and teardown. Important 2: Must cover mocha.js.
- [ ] [wip] Revisit some technical debts (PR feedback notes, possible filenames change, internal TODO) [#2573, #2575, #2577]
- [ ] Dependency Injection - [InversifyJS](https://github.com/inversify/InversifyJS) / [di4js](https://github.com/gedbac/di4js)
- [ ] [tbd] Allow for running debug mode w/ or w/o some sort of instance reuse
- [ ] [tbd] Add support for running on a specific cloud instance (mostly for local dev usage / debugging)
- [ ] Optimize post-testing idle usage time by introducing a worker-teardown hook into Jest, then use it to kill the associated instance (note: currently, all workers -- and thus cloud instances are kept alive until the very end of the suite).
- [ ] Internal usage/error monitoring system. this is a TBD.

**Comment by nhoughto** [comment]
Would be great to support a way of hooking into geny motion device naming too, we want to create a fallback device cleanup process to cleanup devices whose CI job has completed, but the `Detox-<timestamp>-<jest index>` can't be related back to a CI job, so there is no way of knowing which device is for which job.

As we cant guarantee that the detox device cleanup process will _always_ fire a fallback is required.

**Comment by d4vidi** [comment]
Hey @nhoughto, quite exciting to see interest in this feature so soon (hasn't yet officially launched...)
Detox does take care of cleaning up all used Genymotion-cloud instances. You have to issue a call to `detox.globalInit()` in Jest's global init script, and an equivalent `detox.globalCleanup()` call in Jest's global teardown.

When all of this is mature enough, we will properly document it (as mentioned in the issue' description 😄)

**Comment by nhoughto** [comment]
Yeah I figured it wasn’t quite finished yet, still works quite well.

globalCleanup is useful, but can’t solve the entire problem as depending on what CI system you use you can’t ensure the cleanup is actually fired, the job may fail, process run out of memory etc and never run.. would be useful be able to name the device so a cleanup task could relate device to ci job. 👍🏼

**Comment by nhoughto** [comment]
Yeah adding tags to the devices would be better, failing that jamming context into the only place we have would work too =)

**Comment by d4vidi** [comment]
I understand of course :)

**Comment by kg-currenxie** [comment]
Hi. Happy to see support for this, as we're struggling with stability running Android in CircleCI (weird out-of-memory issues, even tho there should be enough)

Just tested Genycloud and to my surprise it works incredibly well! 5/5 re-runs passed, and a bit faster than our current runs.
Though I tested with the cli locally, and not in CI, but it looks like it would work the same.

PS: I've tested both with `--workers` 1 and 2. Both working 👌 

--- 

My only question is, since the issue is still [Open], and my testing was successful, what is left to do here? Could we expect some stuff _not_ working? 🤔

**Comment by noomorph** [comment]
Hi @kg-currenxie, thanks for the cheering feedback. As for your question:

1. First of all, we will release the official Genymotion SaaS support as a part of Detox 20 in the upcoming days.
2. There are minor issues in some advanced scenarios, e.g., #3573, and #3655. These are rather minor inconveniences, but we're not satisfied with their existence, of course, and we are committed to fixing them within a month or two at most.
3. Last but not least, we want to improve init performance dramatically for projects with numerous test suites by dropping unnecessary app re-installation where possible (this is not Genymotion SaaS-specific, but this is exactly an entry point to fix the aforementioned minor issues).

**Comment by kg-currenxie** [comment]
@noomorph sounds awesome :D Then we're happy to make the switch <3

**Comment by noomorph** [comment]
The `next` (20) version and its docs are basically ready, so you can try to install `detox@20.0.13-prerelease.0` and follow the [Migration Guide](https://wix.github.io/Detox/docs/next/guide/migration#200) and [Genymotion SaaS guide](https://wix.github.io/Detox/docs/next/guide/genymotion-saas), to prepare your project in a Git branch towards the upcoming release.

**Comment by kg-currenxie** [comment]
Just to confirm; does it still support recording videos? :) Didn't see any mention of it in the caveats section of the migration guide :)

**Comment by noomorph** [comment]
> Just to confirm; does it still support recording videos? :) Didn't see any mention of it in the caveats section of the migration guide :)

Is there any specific issue you have spotted?

Well, there are no video-specific changes in Detox 20. The support remains basic - 1 video per test - like it always has been since Detox 8. Advanced video cutting is planned but without an ETA at the moment.

**Comment by kg-currenxie** [comment]
No, no issues, just didn't have time to test video recording until my free Genycloud minutes ran out 😅 Thanks

**Comment by kg-currenxie** [comment]
I'll report back when we have done the v20 prerelease migration (next week)

**Comment by kg-currenxie** [comment]
So far, fairly good! After the upgrades (now v20.0.1), I've tried running it 5 times in a row so far.
Same specs, same config, same command, right after each other. (2 workers)
Results in seconds:
- ✅ 432s
- ✅ 439s
- ✅ 479s
- ✅ 516s (17% slower than fastest run, but still fast imo)
- ❌ One failed run

It's not quite clear what the failure is. Some kind of network timeout I think. I'll report here, but it might be more on Geny's side. I'll give their support this issue.

Log (trace level, 65k lines):
https://gist.githubusercontent.com/kg-currenxie/8e1f428138920489c7afcb5c1096c820/raw/5dfdcfa1d0fb3ab0236c4af2e43cbe61ef4ebacc/gistfile1.txt

Please cmd+f for `failed` in it :)

**Comment by d4vidi** [comment]
@kg-currenxie thanks, but it's hard to tell from the log. Would be great if you could share your `detox.trace.json` artifact (https://wix.github.io/Detox/docs/api/logger/#detoxtracejson)

**Comment by kg-currenxie** [comment]
@d4vidi i'll provide the json trace next time it fails :) i didnt have `--record-logs` on :/ 

For now, i'll give you guys the ones that have big time differences:

399 seconds: 
[detox.trace.json.zip](https://github.com/wix/Detox/files/10062406/detox.trace.json.zip)

513 seconds:
[detox.trace.json.zip](https://github.com/wix/Detox/files/10062407/detox.trace.json.zip)

28% time difference, altho, still fast enough for our liking :) would just be interesting to gain any insights on why

**Comment by d4vidi** [comment]
Hey @kg-currenxie! This specific problem doesn't seem to be related to Genymotion, so let's follow up on it in #3722 in order to keep things clean over there 🙏🏻
